<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Web-content" data-toc-modified-id="Web-content-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Web content</a></span><ul class="toc-item"><li><span><a href="#HTML-+-CSS" data-toc-modified-id="HTML-+-CSS-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>HTML + CSS</a></span></li></ul></li><li><span><a href="#Requests-Library" data-toc-modified-id="Requests-Library-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Requests Library</a></span></li><li><span><a href="#Get-our-first-useful-data." data-toc-modified-id="Get-our-first-useful-data.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Get our first useful data.</a></span></li></ul></div>

# Web content

What is the biggest source if data? Obviously Internet. The biggest source of information we can't live without.  
Just imagine yourself the day when we lose Internet. What will you do? Nothing, right? No more Netflix, no more Wikipedia, no more Spotify. Most of the entertainment is web based nowadays. At the same time, all of the bank transactions are done via Internet. So, you won't be able to pay via your phone (NFC technology), via your Contactless card, or even simple card. The only way to pay - via cash. That sucks.

For data people internet is not just a place for entertainment, but the source of data. For some companies Internet is the only mean of existance. They use data from web, aggregate it and sell it back. Smart dudes :D

Web scraping tools are designed to extract data from websites. These tools are useful for those who wants to get data from the Internet. Web scraping is a technology that allows you to automize data collection: instead of opening windows and copy pasting the content you have to write a short script and collect any volume of information. These tools allow you to manually or automatically retrieve new or updated data and save it for later use. For example, using web scraping tools you can extract information about products and prices from online stores.

Possible scenarios for using web scraping tools:

* Data collection for marketing research
* Contact information extraction (email addresses, phone numbers, etc.) from different sites
* StackOverFlow solution retrieval (or other similar resource) for offline access
* Job offer real-time collection
* Price tracking

Depending on complexity and time expenses web pages can be split by 2 main groups:

* API based websites
* HTML based websites

For those problems you need completely different approach. For API based websites solution can be found and data can be collected within several minutes, while for HTML based websites it usually takes few hours to collect all nessesary information.

As a real-life example you can find [here](https://github.com/andreybavt/apart_facile).

How do we get this data?

## HTML + CSS

What can we see on the website?

1. HTML code. Defines the structure of webpage. Simply it is a placeholder for every element on the page, such as links, images, text, videoplayers, etc. Has a very specific structured syntax.
2. Image. 
3. Audio. Not popular type of objects. Usually cashed to your laptop when you load the page, therefore easily accessible.
4. Video. Popular nowadays. Mostly streamed to the webpage, therefore uploaded as encoded objects in chunks - short pieces of video (therefore, difficult to download).
5. CSS code. Defines the styling of objects on webpage. Color, font, size, etc. are defined with CSS.
6. JavaScript code. Operates with dynamic objects on the webpage, such as ads or streamed data. Usually connects webpage with database.
7. Fonts. Just like an audio, cashed to your laptop. 
8. Application data. Whenever you see pdf files being displayed on the page - it is Adobe/Fine/etc. reader that is connected to the webpage to display the file.

The list above is not full, but it describes most of objects you can see on the page. What kind of information we usually care about? Text! Text is easy to read and process. Much easier than anything else. So, our main point of interest is HTML and CSS code.

<img src="https://camo.githubusercontent.com/2f5fe0786f073bf23b62ada3f864327642516dc9/68747470733a2f2f73332d65752d776573742d312e616d617a6f6e6177732e636f6d2f69682d6d6174657269616c732f75706c6f6164732f646174612d7374617469632f696d616765732f68746d6c2d6578616d706c652e706e67">

Here you can see [this webpage]('https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/documents/the-html5-breakfast-site.html').

What does this page looks like?

```html
<!DOCTYPE html>
<html>
    <head>
        <title>The HTML5 Breakfast Site</title>
        <meta charset="utf-8"/>
        <style>
            body {
                background-color: limegreen;
                margin: 40px;
                } 
        </style>
    </head>
    <body>
        <div id="container">
            <nav id="topnav">
                <a href="https://www.ironhack.com" target="_blank">HOME</a> 
                <a href="https://www.ironhack.com/en/team" target="_blank">ABOUT US</a> 
                <a href="https://www.ironhack.com/en/contact" target="_blank">CONTACT US</a>
            </nav>
            <section id="content">
                <h1>The Ironhack Breakfast Place</h1>
                <p>Here you will find all sorts of delicious treats</p>
                <figure>
                    <img alt="healthy breakfast" src="breakfast.jpg" width="400"/>
                    <figcaption>CC Image courtesy of Ruth Hartnup on Flickr</figcaption>
                </figure>
            </section>
            <footer>
                <p class="love">Made with love by Ironhack</p>
            </footer>
        </div>
    </body>
</html>
```

This is the Source Code of the link above. What do we see here? We see structured code with each element being located in specific order. Let's go step by step.
1. 
```html
<!DOCTYPE html>
```
HTML pages can be written in different syntax models. The most common one is HTML. Second most populat is XML.
2. ```html
<html>
    ...
</html>
```
Defines the webpage. Everything inside is the body of your page.
3. ```html
<head>
    ...
</head>
```
Defines the header. This part is invisible on the page. Here you define meta data of webpage, keywords, styling, etc.
4. ```html
<body>
    ...
</body>
```
Defines the webpage itself. Everything defined here will be visible for end user. 
5. ```html
<div>
    ...
</div>
```
Block (container) element. Can be located, sized, parametrized on the page. Every element inside will follow rules defined in `div` block.
6. ```html
<a href='some link'>
    ...
</a>
```
Hyperlink. Has `href` attribute that indicates the target of hyperlink.
7. ```html
<h*>
    ...
</h*>
```
Header. Starts from `h1`, which indicates the biggest font size and top header, goes all the way to `h6` which is the smallest header.
8. ```html
<p>
    ...
</p>
```
Plaintext. 
9. ```html
<img src='some link'>
    ...
</img>
```
Image. Has `src` attribute that indicates the location of image. Width and Heigth can be indicated.



In terms of `css`, we can see that some of the tags have `id` and `class` being specified. 

`id` attribute is a unique identifier of a tag. Usually is applied for very specific pieces of webpage that is updated frequently. `class` attribute is a styling attribute. Tags with the same styling are objects of the same class family. So, if you need to identically format 20 objects on a page you just specify their class to be the same.

For web scrappers, those two attributes are core anchors to detect correct piece of information.  
Please, find [here](https://flukeout.github.io) a tutorial on `css` selectors.

# Requests Library

In [2]:
import requests as r

`requests` library allows us to send requests to any webpage. It provides full control over requests we send, allowing us change proxies, IP-addresses, specifying headers, payloads, etc. Two words: perfect scraping.

In [3]:
r1=r.get('https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/documents/the-html5-breakfast-site.html')
r1

<Response [200]>

In [4]:
type(r1)

requests.models.Response

`get` method allows us send a `GET` request to the webpage. Response is stored in the r1 object of `requests.models.Response` class. As we can see, class displays the status code of the page when the object is printed.

[Here](https://www.restapitutorial.com/httpstatuscodes.html) you can find full list of status codes you can receive from a webpage. Most important are following:
* 1xx - Informational. You won't receive this code while scrapping.
* 2xx - OK. Request was sent correctly and server responded correctly.
* 3xx - Redirect.
* 4xx - Request was sent incorrectly / You are blocked / Page doesn't exist
* 5xx - Request was sent correclty, but server can't responde. 

As you can deduce, 200 is the only accepted.

In [5]:
r2=r.get('https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/documents/forbidden')
r2

<Response [403]>

Sometimes you will receive a bad 403 response. 

In [6]:
r3=r.get('http://google.com')

We all know that few years ago almost all the websites on Internet switched from `http` to `https` protocol. What if I try to access `http` webpage?

In [7]:
r3

<Response [200]>

I receive a 200 status code. Which is cool. But...

In [8]:
r3.history

[<Response [301]>]

In `history` attribute of response object I can see the history of all redirections. So, since `http` is not in use anymore, webpage automatically redirects us to `https` based webpage. We can see it in history - the first link received 301 status code, which is redirect.

Similarly, if I go to `http://fb.com` I will be redirected. Twice.

In [9]:
r3=r.get('http://fb.com')
r3.history

[<Response [302]>, <Response [301]>]

In [12]:
type(r1.status_code)

int

Keeping in mind that we always want to automize processes, let's define a function with basic error handling.

In [14]:
def get_data(url):
    response=r.get(url)
    if response.status_code==200:
        print('alles gut')
    elif response.status_code>400:
        print('error')

In [15]:
get_data('http://fb.com')

alles gut


In [16]:
get_data('https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/documents/forbidden')

error


Seems to be nice, right? In fact, `requests` library already simplified the whole process.

In [17]:
def get_data(url):
    response=r.get(url)
    if response:
        print('alles gut')
    else:
        print('error')

In [18]:
get_data('http://fb.com')

alles gut


In [19]:
get_data('https://s3-eu-west-1.amazonaws.com/ih-materials/uploads/data-static/documents/forbidden')

error


If you use a Response instance in a conditional expression, it will evaluate to `True` if the status code was between 200 and 400, and `False` otherwise.

# Get our first useful data.

Let's collect some data from Allociné. There is a new article about Batman: "*15 ans de Batman Begins : comment Christopher Nolan a sauvé l'Homme Chauve-Souris*".

Most of the webpages are loaded using `GET` request.

In [36]:
responce=r.get('http://www.allocine.fr/article/fichearticle_gen_carticle=18690706.html')

In [28]:
response = r.get('https://www.france24.com/fr/20200617-hidalgo-dati-buzyn-trois-candidates-pour-paris-trois-stratégies')
print(response.content)

b'<HTML><HEAD>\n<TITLE>Access Denied</TITLE>\n</HEAD><BODY>\n<H1>Access Denied</H1>\n \nYou don\'t have permission to access "http&#58;&#47;&#47;www&#46;france24&#46;com&#47;fr&#47;20200617&#45;hidalgo&#45;dati&#45;buzyn&#45;trois&#45;candidates&#45;pour&#45;paris&#45;trois&#45;strat&#37;C3&#37;A9gies" on this server.<P>\nReference&#32;&#35;18&#46;2df31402&#46;1592390642&#46;1e12d267\n</BODY>\n</HTML>\n'
